pkg/providers/k3d/k3d.go

```diff
@@ -0,0 +1,796 @@
+	base := cluster.NewBaseProvider()
+	base.Provider = providerName
+	base.Master = k3dMaster
+	return &K3d{
+		ProviderBase: base,
+		Options: typesk3d.Options{
+			APIPort: k3dAPIPort,
+			Image:   k3dImage,
+		},
+	}
+}
+
+func (p *K3d) GetProviderName() string {
+	return p.Provider
+}
+
+func (p *K3d) GenerateClusterName() string {
+	// must comply with the k3d cluster name rules.
+	p.ContextName = fmt.Sprintf("%s-%s", p.GetProviderName(), p.Name)
+	return p.ContextName
+}
+
+func (p *K3d) CreateK3sCluster() (err error) {
+	return p.InitCluster(p.Options, nil, p.createK3d, p.obtainKubeCfg, p.rollbackK3d)
+}
+
+func (p *K3d) JoinK3sNode() (err error) {
+	return p.JoinNodes(p.joinK3d, p.syncK3d, true, p.rollbackK3d)
+}
+
+func (p *K3d) DeleteK3sCluster(f bool) (err error) {
+	return p.DeleteCluster(f, p.deleteK3d)
+}
+
+func (p *K3d) SSHK3sNode(ip string) error {
+	c := &types.Cluster{
+		Metadata: p.Metadata,
+		Options:  p.Options,
+		Status:   p.Status,
+	}
+	return p.Connect(ip, nil, c, p.k3dStatus, p.isNodeRunning, p.attachNode)
+}
+
+func (p *K3d) IsClusterExist() (bool, []string, error) {
+	ids := make([]string, 0)
+
+	cfg := &k3d.Cluster{
+		Name: p.Name,
+	}
+
+	var (
+		c   *k3d.Cluster
+		err error
+	)
+
+	if c, err = client.ClusterGet(context.Background(), runtimes.SelectedRuntime, cfg); err != nil {
+		// ignore the error.
+		return false, ids, nil
+	}
+
+	for _, n := range c.Nodes {
+		if n.State.Running {
+			ids = append(ids, n.Name)
+		}
+	}
+
+	return len(ids) > 0, ids, nil
+}
+
+func (p *K3d) SetOptions(opt []byte) error {
+	sourceOption := reflect.ValueOf(&p.Options).Elem()
+	option := &typesk3d.Options{}
+	err := json.Unmarshal(opt, option)
+	if err != nil {
+		return err
+	}
+	targetOption := reflect.ValueOf(option).Elem()
+	utils.MergeConfig(sourceOption, targetOption)
+	return nil
+}
+
+func (p *K3d) GetCluster(kubeConfig string) *types.ClusterInfo {
+	c := &types.ClusterInfo{
+		ID:       p.ContextName,
+		Name:     p.Name,
+		Provider: p.GetProviderName(),
+	}
+	if kubeConfig == "" {
+		return c
+	}
+	return p.GetClusterStatus(kubeConfig, c, p.k3dStatus)
+}
+
+func (p *K3d) DescribeCluster(kubeConfig string) *types.ClusterInfo {
+	c := &types.ClusterInfo{
+		Name:     p.Name,
+		Provider: p.GetProviderName(),
+	}
+	return p.Describe(kubeConfig, c, p.k3dStatus)
+}
+
+func (p *K3d) GetProviderOptions(opt []byte) (interface{}, error) {
+	options := &typesk3d.Options{}
+	err := json.Unmarshal(opt, options)
+	return options, err
+}
+
+func (p *K3d) SetConfig(config []byte) error {
+	c, err := p.SetClusterConfig(config)
+	if err != nil {
+		return err
+	}
+	sourceOption := reflect.ValueOf(&p.Options).Elem()
+	b, err := json.Marshal(c.Options)
+	if err != nil {
+		return err
+	}
+	opt := &typesk3d.Options{}
+	err = json.Unmarshal(b, opt)
+	if err != nil {
+		return err
+	}
+	targetOption := reflect.ValueOf(opt).Elem()
+	utils.MergeConfig(sourceOption, targetOption)
+
+	return nil
+}
+
+func (p *K3d) Rollback() error {
```

Comment:
We have removed the provider interface `Rollback`, so remove the function cause no one is using it.